packages/connection/src/class-nonce-handler.php

```diff
@@ -0,0 +1,197 @@
+			 *
+			 * @param int $limit How many old nonces to remove at shutdown.
+			 */
+			if ( apply_filters( 'jetpack_connection_add_nonce_cleanup', $run_cleanup ) ) {
+				add_action( 'shutdown', array( __CLASS__, 'clean_runtime' ) );
+			}
+		} else {
+			$return = false;
+		}
+
+		$wpdb->show_errors( $show_errors );
+
+		static::$nonces_used_this_request[ "$timestamp:$nonce" ] = $return;
+
+		return $return;
+	}
+
+	/**
+	 * Removing all the nonces.
+	 *
+	 * @param int $cutoff_timestamp All nonces added before this timestamp will be removed.
+	 *
+	 * @return true
+	 */
+	public static function clean_all( $cutoff_timestamp = PHP_INT_MAX ) {
+		for ( $i = 0; $i < 1000; ++$i ) {
+			$result = static::delete( static::CLEAN_ALL_LIMIT_PER_BATCH, $cutoff_timestamp );
+
+			if ( ! $result ) {
+				break;
+			}
+		}
+
+		return true;
+	}
+
+	/**
+	 * Clean up the expired nonces on shutdown.
+	 *
+	 * @return bool True if the cleanup query has been run, false if the table is locked.
+	 */
+	public static function clean_runtime() {
+		if ( static::is_table_locked() ) {
+			return false;
+		}
+
+		/**
+		 * Adjust the number of old nonces that are cleaned up at shutdown.
+		 *
+		 * @since 9.0.0
+		 *
+		 * @param int $limit How many old nonces to remove at shutdown.
+		 */
+		$limit = apply_filters( 'jetpack_connection_nonce_cleanup_runtime_limit', static::CLEANUP_RUNTIME_LIMIT );
+
+		static::delete( $limit, time() - static::LIFETIME );
+
+		return true;
+	}
+
+
+	/**
+	 * Delete the nonces.
+	 *
+	 * @param int      $limit How many nonces to delete.
+	 * @param null|int $cutoff_timestamp All nonces added before this timestamp will be removed.
+	 *
+	 * @return int|false Number of removed nonces, or `false` if nothing to remove (or in case of a database error).
+	 */
+	public static function delete( $limit = 10, $cutoff_timestamp = null ) {
+		global $wpdb;
+
+		$ids = $wpdb->get_col(
+			$wpdb->prepare(
+				"SELECT option_id FROM `{$wpdb->options}`"
+				. " WHERE `option_name` >= 'jetpack_nonce_' AND `option_name` < %s"
+				. ' LIMIT %d',
+				'jetpack_nonce_' . $cutoff_timestamp,
+				$limit
+			)
+		);
+
+		if ( ! is_array( $ids ) || ! count( $ids ) ) {
+			// There's either nothing to remove, or there's an error and we can't proceed.
+			return false;
+		}
+
+		$ids_fill = implode( ', ', array_fill( 0, count( $ids ), '%d' ) );
+
+		// The Code Sniffer is unable to understand what's going on...
+		// phpcs:ignore WordPress.DB.PreparedSQL.InterpolatedNotPrepared,WordPress.DB.PreparedSQLPlaceholders.UnfinishedPrepare
+		return $wpdb->query( $wpdb->prepare( "DELETE FROM `{$wpdb->options}` WHERE `option_id` IN ( {$ids_fill} )", $ids ) );
+	}
+
+	/**
+	 * Clean the cached nonces valid during the current request, therefore making them invalid.
+	 *
+	 * @return bool
+	 */
+	public static function invalidate_request_nonces() {
+		static::$nonces_used_this_request = array();
+
+		return true;
+	}
+
+	/**
+	 * Check if the options table is locked.
+	 *
```

Comment:
Worth acknowledging in the docs that use of this is subject to races, but we don't care since it's enough to mostly work around a long-query issue from some other plugin?

Reply:
Yes, good point, I'll add that to the comments.